README.md
<img src="https://raw.githubusercontent.com/ricardonior29/easy_dialog/master/example/assets/topImage.png" height="100" alt="logo"/>

[![pub package](https://img.shields.io/pub/v/easy_dialog.svg?color=orange&logo=flutter&logoColor=white)](https://pub.dartlang.org/packages/easy_dialog)&nbsp; [![GitHub stars](https://img.shields.io/github/stars/ricardonior29/easy_dialog.svg?color=blue&label=Stars&logo=GitHub)](https://github.com/ricardonior29/easy_dialog)&nbsp; [![GitHub](https://img.shields.io/github/license/ricardonior29/easy_dialog.svg?color=darkred&label=License)](https://github.com/ricardonior29/easy_dialog/blob/master/LICENSE) &nbsp; &nbsp; [![LinkedIn](https://img.shields.io/badge/LinkedIn-RN-informational.svg?logo=LinkedIn&color=0077b5)](https://www.linkedin.com/in/ricardo-ni%C3%B1o-113507183/)&nbsp; [![LinkedIn](https://img.shields.io/badge/LinkedIn-JD-informational.svg?logo=LinkedIn&color=0077b5)](https://www.linkedin.com/in/javier-duarte-269b82181/)
<br>[![Open Source Love](https://badges.frapsoft.com/os/v3/open-source.png?v=103)](https://opensource.org/) 

[comment]: # (https://github.com/ellerbrock/open-source-badges/)

---
Easy Dialog package helps you easily create basic or custom dialogs.
For extended  [documentation](https://pub.dartlang.org/documentation/easy_dialog/latest/easy_dialog/easy_dialog-library.html) visit project pub package.

#### Star ⭐  this repo if you like our project. We would love your support.

# 💾 Installation
In the `dependencies:` section of your `pubspec.yaml`, add the following line:

```yaml
dependencies:
  easy_dialog: <latest version>
```

# 📦 Usage

### Import this class

```dart
import 'package:easy_dialog/easy_dialog.dart';
```
### Basic structure
```dart
EasyDialog(
        title: Text("Basic Easy Dialog Title"),
        description: Text("This is a basic dialog")).show(context);
```

# 🎯 Examples

Let start creating simple but beautiful dialogs. 
<br>[Here](https://github.com/ricardonior29/easy_dialog/tree/master/example) you have some basic examples of what can you build and let your creativity fly.

<img src="https://raw.githubusercontent.com/ricardonior29/easy_dialog/master/example/assets/examples_gif.gif" align = "" height = "500" alt="Network">



# 🎨 Customization and Attributes

All customizable [attributes](https://pub.dartlang.org/documentation/easy_dialog/latest/easy_dialog/EasyDialog-class.html) for easy_dialog

<table>
    <th>Attribute Name</th>
    <th>Example Value</th>
    <th>Description</th>
    <tr>
        <td>topImage</td>
        <td>AssetImage("assets/topImage.png")</td>
        <td>Top image for decoration purpose with rounded top corners</td>
    </tr>
    <tr>
        <td>title</td>
        <td>Text("Alert")</td>
        <td>Dialog title, located under topImage if exists</td>
    </tr>
    <tr>
        <td>tittlePadding</td>
        <td>EdgeInsets.only(left: 8.0,right: 8.0)</td>
        <td>Title padding</td>
    </tr>
    <tr>
        <td>description</td>
        <td>Text("Connection error")</td>
        <td>Dialog description, located between title and contentList</td>
    </tr>
    <tr>
        <td>descriptionPadding</td>
        <td>EdgeInsets.only(left: 8.0,right: 8.0)</td>
        <td>Description padding</td>
    </tr>
    <tr>
        <td>contentList</td>
        <td>[<br>
              Text("This is an extra text line"),
              FlatButton(onPressed: () {},<br>
              &nbsp; child: Text("Okay"),),<br>
              ]
        </td>
        <td>Optional widget list, located in order before description</td>
    </tr>
    <tr>
        <td>contentPadding</td>
        <td>EdgeInsets.all(12.0)</td>
        <td>Content padding. It wraps all widgets including title and description</td>
    </tr>
    <tr>
        <td>height</td>
        <td>200</td>
        <td>Dialog height</td>
    </tr>
    <tr>
        <td>width</td>
        <td>250</td>
        <td>Dialog width</td>
    </tr>
    <tr>
        <td>cornerRadius</td>
        <td>15.0</td>
        <td>Dialog corner radius</td>
    </tr>
    <tr>
        <td>fogOpacity</td>
        <td>0.5</td>
        <td>Fog opacity, sets the black fog opacity behind our dialog from 0 to 1</td>
    </tr>
    <tr>
        <td>closeButton</td>
        <td>false</td>
        <td>Close button enable/disable</td>
    </tr>
</table>

# 🧱 How to Contribute
1. Fork it
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -am 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request


# 📃 MIT License

    Copyright (c) 2019 ricardonior29

    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
    
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).

CHANGELOG.md
1.1.1

    Input contentList bug fix. Now contentList combined with title and description shows all widgets.
    Updates example app.
    
1.1.0

    Null safety and example app.
    
1.0.5

    Card color variable.

1.0.4

    Flutter format to lib/easy_dialog.dart. Added an uploader.

1.0.3

    Changes in readme files.

1.0.2

    Clean example dir. Changes in readme files.

1.0.1

    Minor changes-pubspec.yaml.

1.0.0

    Initial Release.



lib/easy_dialog.dart
library easy_dialog;

import 'package:flutter/material.dart';

class EasyDialog {
  final ImageProvider? topImage;
  final Text? title;
  final Text? description;
  final bool closeButton;
  final double height;
  final double width;
  final double fogOpacity;
  final double cornerRadius;
  final Color cardColor;
  List<Widget> _contentList = [];
  List<Widget>? contentList;
  EdgeInsets? contentPadding;
  EdgeInsets? titlePadding;
  EdgeInsets descriptionPadding;
  CrossAxisAlignment contentListAlignment;

  EasyDialog(
      {Key? key,
      this.topImage,
      this.title,
      this.description,
      this.closeButton = true,
      this.height = 140,
      this.width = 300,
      this.cornerRadius = 8.0,
      this.fogOpacity = 0.37,
      this.cardColor = const Color.fromRGBO(240, 240, 240, 1.0),
      this.contentList,
      this.contentPadding = const EdgeInsets.fromLTRB(17.5, 12.0, 17.5, 13.0),
      this.descriptionPadding = const EdgeInsets.all(0.0),
      this.titlePadding = const EdgeInsets.only(bottom: 12.0),
      this.contentListAlignment = CrossAxisAlignment.center})
      : assert(fogOpacity >= 0 && fogOpacity <= 1.0);

  insertByIndex(EdgeInsets? padding, Widget? child, int index) {
    _contentList.insert(
        index,
        Container(
          padding: padding,
          child: child,
          alignment: Alignment.center,
        ));
  }

  show(BuildContext context) {
    ClipRRect? image;
    if (topImage != null) {
      image = ClipRRect(
        child: new Image(image: topImage!),
        borderRadius: BorderRadius.only(
            topLeft: Radius.circular(cornerRadius),
            topRight: Radius.circular(cornerRadius)),
      );
    }

    if (title != null && description != null) {
      insertByIndex(titlePadding, title, 0);
      insertByIndex(descriptionPadding, description, 1);
    }
    if (title != null && description == null) {
      insertByIndex(titlePadding, title, 0);
    }
    if (description != null && title == null) {
      insertByIndex(descriptionPadding, description, 0);
    }

    contentList?.forEach((element) {
      insertByIndex(EdgeInsets.zero, element, _contentList.length);
    });

    return showDialog(
      context: context,
      barrierDismissible: true,
      builder: (BuildContext context) {
        return Material(
          color: Colors.transparent,
          child: Container(
            color: Color.fromRGBO(0, 0, 0, fogOpacity),
            child: new Center(
              child: new Container(
                  decoration: BoxDecoration(
                    boxShadow: <BoxShadow>[
                      new BoxShadow(
                        spreadRadius: 1.0,
                        color: Colors.black54,
                        offset: new Offset(5.0, 5.0),
                        blurRadius: 30.0,
                      )
                    ],
                    borderRadius:
                        BorderRadius.all(Radius.circular(cornerRadius)),
                    color: cardColor,
                  ),
                  height: height,
                  width: width,
                  child: new Column(
                    children: <Widget>[
                      Stack(
                        children: <Widget>[
                          image == null
                              ? Container(
                                  alignment: Alignment.center,
                                  height: height,
                                  padding: contentPadding,
                                  child: Column(
                                    crossAxisAlignment: contentListAlignment,
                                    mainAxisAlignment: MainAxisAlignment.center,
                                    children: _contentList,
                                  ),
                                )
                              : image,
                          closeButton == true
                              ? Container(
                                  margin: EdgeInsets.only(top: 8),
                                  height: 26,
                                  alignment: Alignment.topRight,
                                  child: FloatingActionButton(
                                    onPressed: () {
                                      Navigator.pop(context);
                                    },
                                    child: Icon(
                                      Icons.close,
                                      color: Colors.black87,
                                      size: 19,
                                    ),
                                    backgroundColor:
                                        Color.fromRGBO(240, 240, 240, 0.8),
                                    elevation: 2,
                                  ),
                                )
                              : Container(),
                        ],
                      ),
                      image == null
                          ? Container()
                          : Expanded(
                              child: Container(
                                padding: contentPadding,
                                child: Column(
                                  crossAxisAlignment: contentListAlignment,
                                  mainAxisAlignment: MainAxisAlignment.end,
                                  children: _contentList,
                                ),
                              ),
                            ),
                    ],
                  )),
            ),
          ),
        );
      },
    );
  }
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"example1", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:easy_dialog/easy_dialog.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Easy Dialog Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Easy Dialog Examples'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  // ignore: library_private_types_in_public_api
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // Easy Dialog without using parameters
  void _emptyEasyDialog() {
    EasyDialog().show(context);
  }

  // Easy Dialog using title and description
  void _basicEasyDialog() {
    EasyDialog(
        title: const Text(
          "Basic Easy Dialog Title",
          style: TextStyle(fontWeight: FontWeight.bold),
          textScaleFactor: 1.2,
        ),
        description: const Text(
          "This is a basic dialog. Easy Dialog helps you easily create basic or custom dialogs.",
          textScaleFactor: 1.1,
          textAlign: TextAlign.center,
        )).show(context);
  }

  // Easy Dialog using title, description, topImage and height
  void _basicImageEasyDialog() {
    EasyDialog(
      title: const Text(
        "Basic Easy Dialog Title",
        style: TextStyle(fontWeight: FontWeight.bold),
        textScaleFactor: 1.2,
      ),
      description: const Text(
        "This is a basic dialog. Easy Dialog helps you easily create basic or custom dialogs.",
        textScaleFactor: 1.1,
        textAlign: TextAlign.center,
      ),
      topImage: const NetworkImage(
          "https://raw.githubusercontent.com/ricardonior29/easy_dialog/master/example/assets/topImageblack.png"),
      height: 180,
    ).show(context);
  }

  // Easy Dialog using title, description, topImage, height and contentList
  void _basicContentEasyDialog() {
    EasyDialog(
        title: const Text(
          "Basic Easy Dialog Title",
          style: TextStyle(fontWeight: FontWeight.bold),
          textScaleFactor: 1.2,
        ),
        description: const Text(
          "This is a basic dialog. Easy Dialog helps you easily create basic or custom dialogs.",
          textScaleFactor: 1.1,
          textAlign: TextAlign.center,
        ),
        topImage: const NetworkImage(
            "https://raw.githubusercontent.com/ricardonior29/easy_dialog/master/example/assets/topImageblack.png"),
        height: 220,
        contentList: [
          Row(
            mainAxisAlignment: MainAxisAlignment.end,
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.only(top: 8.0),
                child: TextButton(
                  style: ButtonStyle(
                    foregroundColor:
                        MaterialStateProperty.all(Colors.lightBlue),
                  ),
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                  child: const Text(
                    "Accept",
                    textScaleFactor: 1.2,
                  ),
                ),
              ),
              Padding(
                padding: const EdgeInsets.only(top: 8.0),
                child: TextButton(
                  style: ButtonStyle(
                    foregroundColor:
                        MaterialStateProperty.all(Colors.lightBlue),
                  ),
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                  child: const Text(
                    "Cancel",
                    textScaleFactor: 1.2,
                  ),
                ),
              ),
            ],
          )
        ]).show(context);
  }

  void _customButtonEasyDialog() {
    EasyDialog(
        closeButton: false,
        cornerRadius: 10.0,
        fogOpacity: 0.1,
        width: 280,
        height: 230,
        title: const Text(
          "Custom Easy Dialog Title",
          style: TextStyle(fontWeight: FontWeight.bold),
          textScaleFactor: 1.2,
        ),
        descriptionPadding:
            const EdgeInsets.only(left: 17.5, right: 17.5, bottom: 15.0),
        description: const Text(
          "This is a custom dialog. Easy Dialog helps you easily create basic or custom dialogs.",
          textScaleFactor: 1.1,
          textAlign: TextAlign.center,
        ),
        topImage: const NetworkImage(
            "https://raw.githubusercontent.com/ricardonior29/easy_dialog/master/example/assets/topImageblack.png"),
        contentPadding:
            const EdgeInsets.only(top: 12.0), // Needed for the button design
        contentList: [
          Container(
            width: double.infinity,
            decoration: const BoxDecoration(
                color: Colors.greenAccent,
                borderRadius: BorderRadius.only(
                    bottomLeft: Radius.circular(10.0),
                    bottomRight: Radius.circular(10.0))),
            child: TextButton(
              onPressed: () {
                Navigator.of(context).pop();
              },
              style: ButtonStyle(
                foregroundColor: MaterialStateProperty.all(Colors.black),
              ),
              child: const Text(
                "Okay",
                textScaleFactor: 1.3,
              ),
            ),
          ),
        ]).show(context);
  }

  void _customRateEasyDialog() {
    EasyDialog(
        cornerRadius: 15.0,
        fogOpacity: 0.1,
        width: 280,
        height: 230,
        contentPadding:
            const EdgeInsets.only(top: 12.0), // Needed for the button design
        contentList: [
          Expanded(
            flex: 1,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.start,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: <Widget>[
                const Padding(padding: EdgeInsets.only(left: 30.0)),
                const Text(
                  "Rate",
                  style: TextStyle(fontWeight: FontWeight.bold),
                  textScaleFactor: 1.3,
                ),
                const Padding(padding: EdgeInsets.only(left: 10.0)),
                Row(
                  mainAxisSize: MainAxisSize.min,
                  children: List.generate(5, (index) {
                    return Icon(
                      index < 3 ? Icons.star : Icons.star_border,
                      size: 30.0,
                      color: Colors.orange,
                    );
                  }),
                )
              ],
            ),
          ),
          Expanded(
              flex: 3,
              child: Padding(
                padding: const EdgeInsets.all(10.0),
                child: TextFormField(
                  maxLines: 5,
                  decoration: const InputDecoration(
                    border: InputBorder.none,
                    hintText: "Add review",
                  ),
                ),
              )),
          Container(
            width: double.infinity,
            decoration: const BoxDecoration(
                color: Colors.greenAccent,
                borderRadius: BorderRadius.only(
                    bottomLeft: Radius.circular(10.0),
                    bottomRight: Radius.circular(10.0))),
            child: TextButton(
              onPressed: () {
                Navigator.of(context).pop();
              },
              style: ButtonStyle(
                foregroundColor: MaterialStateProperty.all(Colors.black),
              ),
              child: const Text(
                "Rate",
                textScaleFactor: 1.3,
              ),
            ),
          ),
        ]).show(context);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _emptyEasyDialog,
              style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.blue),
                foregroundColor: MaterialStateProperty.all(Colors.white),
              ),
              child: const Text(
                "Empty Dialog",
                textScaleFactor: 1.2,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _basicEasyDialog,
              style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.blue),
                foregroundColor: MaterialStateProperty.all(Colors.white),
              ),
              child: const Text(
                "Basic 1 Dialog",
                textScaleFactor: 1.2,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _basicImageEasyDialog,
              style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.blue),
                foregroundColor: MaterialStateProperty.all(Colors.white),
              ),
              child: const Text(
                "Basic 2 Dialog",
                textScaleFactor: 1.2,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _basicContentEasyDialog,
              style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.blue),
                foregroundColor: MaterialStateProperty.all(Colors.white),
              ),
              child: const Text(
                "Basic 3 Dialog",
                textScaleFactor: 1.2,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _customButtonEasyDialog,
              style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.blue),
                foregroundColor: MaterialStateProperty.all(Colors.white),
              ),
              child: const Text(
                "Custom 1 Dialog",
                textScaleFactor: 1.2,
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _customRateEasyDialog,
              style: ButtonStyle(
                backgroundColor: MaterialStateProperty.all(Colors.blue),
                foregroundColor: MaterialStateProperty.all(Colors.white),
              ),
              child: const Text(
                "Custom 2 Dialog",
                textScaleFactor: 1.2,
              ),
            ),
          ),
        ],
      )),
      // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
